### ISSUE TITLE
Bug: Incorrect (3, 1) Riemann Tensor computation in PyTorch backend due to PyTorch autodiff jacobian `create_graph` being turned off

### ISSUE THREAD

**Issue by yangshengaa** [geomstats/geomstats#1928]
### Describe the bug

To compute the (3, 1) Riemann tensor, we need to differentiate through the Christoffel symbol of the second kind, which in turn is computed by differentiating through the metric. Both diffferentiations rely on autodiff in `pytorch.autograd.jacobian` as inherited from [geomstats._backend.pytorch.autodiff](https://github.com/geomstats/geomstats/blob/0102ceaeac84c39a9806cb1bd2853dbbbc134e48/geomstats/_backend/pytorch/autodiff.py#L90), but the `create_graph` stays `False` by default. This means the second time we call autodiff, no gradient information is tracked and thus the first two terms in Riemann tensor are zeros regardless. 

We could expose the `create_graph` argument to users with default still set to `False`, and just change locally how we compute the Riemann tensor without affecting other codes. 

Similar issues has been brought up, see [Bug: Jacobian of Jacobian in pytorch produces null matrices #1801](https://github.com/geomstats/geomstats/issues/1801) also. 

### Steps/Code to Reproduce

Below demonstrate that recursive calling to pytroch jacobian yields zero. 

```
# treat a single hidden layer neural net feature map as an immersion 
d = 3  # input dim 
w = 10 # hidden dim 
W = torch.randn((w, d))
nl = nn.Sigmoid() # Sigmoid nonlinearity

class NNFeatMap(ImmersedSet):
    def __init__(self, equip=True):
        super().__init__(dim=d, equip=equip)

    def immersion(self, point):
        res = nl(point @ W.T)
        return res

    def _define_embedding_space(self):
        return Euclidean(dim=w)

gs_model = NNFeatMap(equip=False)
gs_model.equip_with_metric(PullbackMetric)

# random input 
X = torch.randn(4, d)
jacobian_christoffels = gs.autodiff.jacobian_vec(gs_model.metric.christoffels)(X)
```

With Sigmoid nonlinearity, `jacobian_christoffels` should not be zero under random input, but in fact they are all zero. This is due to the mechnism above. 

This suggests the first two terms in Riemann tensor computation is incorrect, since they call exactly `gs.autodiff.jacobian_vec(self.metric.christoffels)(base_point)`.

### Expected Behaviour

Expected nonzero values.

### Actual Behaviour

Only zeros are thrown. 

### Your environment

```shell
geomstats==2.7.0

Mac OS
```

**Comment by luisfpereira** [comment]
Dear @yangshengaa, thanks for bringing attention to this.

Do you know what is the impact of `create_graph=True`, i.e. why does `pytorch` have such an option instead of simply defaulting it to `True`? In terms of usability, it is a bit dangerous imo.

In any case, would you be willing to add the flag `create_graph` to our backend? Notice it has also to be added to `numpy` and `autograd`, with the information in the docstring that it is ignored. Otherwise, I'll try to do it asap.

**Comment by yangshengaa** [comment]
My understanding is that they turn it off for computational concerns, since most optimization in Deep learning is first order. After getting the Jacobian we are done. Empirically, `create_graph=True` and `create_graph=False` have substantially different memory consumptions and runtimes, with the first option much more unfriendly. 

I would like to take your expertise on this (I am not an expert in system design), since if we change `create_graph=True` throughout, it incurs unnecessary computational costs for users. I think it make sense only to enable this for recursive calls? But of course this would mean quite some coding. But if this is consistent with `numpy` and `autograd` then we could just do that. 

If you agree, I am happy to open a pull request to just add the flag there without changing other stuffs.

**Comment by luisfpereira** [comment]
Makes sense. I'm just wondering why this does not seem to be a concern to `jax` (see e.g. [example](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#jacobians-and-hessians-using-jacfwd-and-jacrev) and [signature](https://jax.readthedocs.io/en/latest/_autosummary/jax.jacfwd.html)). Anyway, it does not really matter for our purposes, I'm just mentioning this because we want to bring  in `jax` in the future.

I agree users should not incur in additional costs, so I agree with you about only enabling it for recursive calls (with the flag `create_graph`; defaults to `False`). The main challenge may be knowing when creating the first function if it will ever be differentiated. But currently, there's not a lot of places where we use it (at first sight, `ImmersedSet` seems to be the only place where we will need to add the flag).

Regarding the other backends, `numpy` does not support `autodiff`, and `autograd` has the `create_graph=True` behavior by design (cannot be changed). This means `create_graph` will not affect them, though we have to have it in the function signature for compatibility (with docstrings mentioning it is ignored).

This is not perfect from an API perspective, but given the differences in the backend behavior, I can't see a way out (maybe in the future we can come up with something nicer).

From my side, you have green light to open a PR adding a flag `create_graph` to the backends (all of them, though only impacting `pytorch`) and updating the calls of the function in the library (probably only need to change `ImmersedSet`). Thanks again!

**Comment by yangshengaa** [comment]
I am a bit puzzled in how to submit a PR for the moment. Are you able to do this? Thanks!
